### PR TITLE
Feature/gateway

### PR DESCRIPTION
**Pull request by sarvesh-ost**
1. Created gateway setup and co-gateway setup contracts. 
2. Added message bus code hash in gateway link intent declaration

### CODE REVIEW

File: contracts/gateway/GatewayBase.sol [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+    );
+
+    bytes32 constant REDEEM_TYPEHASH = keccak256(
+        abi.encode(
+            "Redeem(uint256 amount,address beneficiary,MessageBus.Message message)"
+        )
+    );
+    bytes32 constant GATEWAY_LINK_TYPEHASH = keccak256(
+        abi.encode(
+            "GatewayLink(bytes32 messageHash,MessageBus.Message message)"
```
Comment (deepesh-kn) [inline contracts/gateway/GatewayBase.sol]: Add `Message` struct as specified in EIP-712 for `STAKE_TYPEHASH`, `REDEEM_TYPEHASH `, `GATEWAY_LINK_TYPEHASH `

File: contracts/gateway/GatewayBase.sol [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+        )
+    );
+
+    /* constants */
```
Comment (deepesh-kn) [inline contracts/gateway/GatewayBase.sol]: A new line after this comment

File: contracts/gateway/GatewayBase.sol [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+    /** Position of message bus in the storage */
+    uint8 constant MESSAGE_BOX_OFFSET = 1;
+    /** Penalty in bounty amount percentage charged to staker on revert staking
```
Comment (deepesh-kn) [inline contracts/gateway/GatewayBase.sol]: New line after L64.

If the comment doesn't fit in one line then split it in to multiline (i.e comment will start from L66)

File: contracts/gateway/GatewayBase.sol [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+     */
+    uint8 constant REVOCATION_PENALTY = 150;
+
+    /** Specifies if the Gateway and CoGateway contracts are linked. */
+    bool public linked;
+
+    /**
+     * Message box.
+     * @dev keep this is at location 1, in case this is changed then update
+     *      constant OUTBOX_OFFSET accordingly.
```
Comment (deepesh-kn) [inline contracts/gateway/GatewayBase.sol]: rename`OUTBOX_OFFSET ` to `MESSAGE_BOX_OFFSET`

File: contracts/gateway/GatewayBase.sol [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+     */
+    MessageBus.MessageBox messageBox;
+
```
Comment (deepesh-kn) [inline contracts/gateway/GatewayBase.sol]: remove extra newline

File: contracts/gateway/GatewayBase.sol [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+
+    /** Specifies if the Gateway is deactivated for any new staking process. */
```
Comment (deepesh-kn) [inline contracts/gateway/GatewayBase.sol]: /** Specifies if the Gateway is deactivated for any new process initiation. */

File: contracts/gateway/GatewayBase.sol [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+    bool public deactivated;
+
+    /** Organisation address. */
+    address public organisation;
+
+    /** amount of ERC20 which is staked by facilitator. */
```
Comment (deepesh-kn) [inline contracts/gateway/GatewayBase.sol]: Can we just call it BaseToken ? (ie. ERC on Gateway and BaseToken in CoGateway)

File: contracts/gateway/GatewayBase.sol [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+    uint256 public bounty;
+
```
Comment (deepesh-kn) [inline contracts/gateway/GatewayBase.sol]: Remove extra line

File: contracts/gateway/GatewayBase.sol [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+
+    /** address of core contract. */
+    CoreInterface public core;
+
+    /** path to prove merkle account proof for CoGateway contract. */
```
Comment (deepesh-kn) [inline contracts/gateway/GatewayBase.sol]: Gateway/CoGateway

File: contracts/gateway/GatewayBase.sol [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+
+    /**
+     * Maps address to ActiveProcess object.
+     *
+     * Once the minting or redeem process is started the corresponding
+     * message hash is stored in ActiveProcess against the staker/redeemer
+     * address. This is used to restrict simultaneous/multiple minting and
+     * redeem for a particular address. This is also used to determine the
+     * nonce of the particular address. Refer getNonce for the details.
+     */
```
Comment (deepesh-kn) [inline contracts/gateway/GatewayBase.sol]: As its now in base, update the comment to be more generic.

File: contracts/gateway/GatewayBase.sol [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+    mapping(address /*address*/ => ActiveProcess) activeProcess;
+
+    /* internal variables */
+
+    /** address of message bus used to fetch codehash during gateway linking */
+    address internal messageBus;
+
```
Comment (deepesh-kn) [inline contracts/gateway/GatewayBase.sol]: Remove new line

File: contracts/gateway/Gateway.sol [hunk 12]
```diff
@@ -15,142 +15,22 @@ pragma solidity ^0.4.23;
 // limitations under the License.
 //
 // ----------------------------------------------------------------------------
-// Origin Chain: Gateway Contract
+// Origin Chain: GatewaySetup Contract
```
Comment (deepesh-kn) [inline contracts/gateway/Gateway.sol]: Gateway

File: contracts/gateway/EIP20Gateway.sol [hunk 13]
```diff
@@ -0,0 +1,1254 @@
...
+// ----------------------------------------------------------------------------
+
+
+/*
+
+           Origin chain      |       Auxiliary chain
+-------------------------------------------------------------------------------
+           Gateway - - - - - - - - - - - CoGateway
+-------------------------------------------------------------------------------
+1. GatewayLinking:
```
Comment (deepesh-kn) [inline contracts/gateway/EIP20Gateway.sol]: GatewayLinking,  should be moved to Gateway.sol and CoGateway.sol

File: contracts/gateway/GatewayBase.sol [hunk 14]
```diff
@@ -0,0 +1,395 @@
...
+ *  @notice GatewayBase contains general purpose functions shared between
+ *  gateway and co-gateway contract.
+ */
+contract GatewayBase {
+
+    using SafeMath for uint256;
+    /** Emitted whenever a Gateway/CoGateway contract is proven.
+     *	wasAlreadyProved parameter differentiates between first call and replay
+     *  call of proveGateway method for same block height
+     */
```
[the remark this answers is not in the document]
Comment (sarvesh-ost) [reply contracts/gateway/GatewayBase.sol]: Done

File: contracts/gateway/GatewayBase.sol (hunk 1 above)
Comment (sarvesh-ost) [reply contracts/gateway/GatewayBase.sol]: Created issue for this.
#347

File: contracts/gateway/GatewayBase.sol (hunk 2 above)
Comment (sarvesh-ost) [reply contracts/gateway/GatewayBase.sol]: Done

File: contracts/gateway/GatewayBase.sol (hunk 3 above)
Comment (sarvesh-ost) [reply contracts/gateway/GatewayBase.sol]: Done

File: contracts/gateway/GatewayBase.sol (hunk 4 above)
Comment (sarvesh-ost) [reply contracts/gateway/GatewayBase.sol]: Done

File: contracts/gateway/GatewayBase.sol (hunk 9 above)
Comment (sarvesh-ost) [reply contracts/gateway/GatewayBase.sol]: Done

File: contracts/gateway/GatewayBase.sol (hunk 10 above)
Comment (sarvesh-ost) [reply contracts/gateway/GatewayBase.sol]: Done

File: contracts/gateway/GatewayBase.sol (hunk 11 above)
Comment (sarvesh-ost) [reply contracts/gateway/GatewayBase.sol]: Done

File: contracts/gateway/Gateway.sol (hunk 12 above)
Comment (sarvesh-ost) [reply contracts/gateway/Gateway.sol]: Done

File: contracts/gateway/EIP20Gateway.sol (hunk 13 above)
Comment (sarvesh-ost) [reply contracts/gateway/EIP20Gateway.sol]: Let's discuss this.

Comment (sarvesh-ost) [reply contracts/gateway/EIP20Gateway.sol]: Done

**Pull request merged**